File: app/src/main/java/com/pauloavelar/inventory/ProductAdapter.java
package com.pauloavelar.inventory;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import java.util.ArrayList;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ViewHolder> {

    public interface OnInteraction {
        void onClickEdit(String product);
        void onClickDelete(String product);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView product;
        ImageButton edit, delete;

        public ViewHolder(View itemView) {
            super(itemView);
            product = (TextView)    itemView.findViewById(R.id.product_name);
            edit    = (ImageButton) itemView.findViewById(R.id.button_product_edit);
            delete  = (ImageButton) itemView.findViewById(R.id.button_product_delete);
        }

        public void bind(final String productName, final OnInteraction listener) {
            product.setText(productName);
            edit.setOnClickListener(new View.OnClickListener() {
                @Override public void onClick(View view) {
                    listener.onClickEdit(productName);
                }
            });
            delete.setOnClickListener(new View.OnClickListener() {
                @Override public void onClick(View view) {
                    listener.onClickDelete(productName);
                }
            });
        }
    }

    private ArrayList<String> mItems;
    private OnInteraction mListener;

    public void clearAll() {
        if (mItems != null) mItems.clear();
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View rowView = LayoutInflater.from(parent.getContext())
                                     .inflate(R.layout.product_row, parent, false);
        return new ViewHolder(rowView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        String item = mItems.get(position);
        holder.bind(item, mListener);
    }

    @Override
    public long getItemId(int i) { return 0; }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public ProductAdapter(OnInteraction listener) {
        mListener = listener;
    }

    public void setItems(ArrayList<String> items) {
        mItems = items;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/pauloavelar/inventory/ui/component/ProductAdapter.kt
package com.pauloavelar.inventory.ui.component

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.pauloavelar.inventory.R
import com.pauloavelar.inventory.model.Product

class ProductAdapter(
    private val listener: OnItemInteraction
) : RecyclerView.Adapter<ProductAdapter.ViewHolder>() {

    private var mItems: List<Product> = emptyList()

    fun setItems(items: List<Product>) {
        mItems = items
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val rowView = LayoutInflater.from(parent.context)
            .inflate(R.layout.product_row, parent, false)
        return ViewHolder(rowView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(mItems[position], listener)
    }

    override fun getItemId(i: Int): Long = i.toLong()

    override fun getItemCount(): Int = mItems.size

    interface OnItemInteraction {
        fun onClickEdit(product: Product)
        fun onClickDelete(product: Product)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var txtProductName: TextView = view.findViewById(R.id.product_name)
        var btnEdit: ImageButton = view.findViewById(R.id.button_product_edit)
        var btnDelete: ImageButton = view.findViewById(R.id.button_product_delete)

        fun bind(product: Product, listener: OnItemInteraction) {
            txtProductName.text = product.name
            btnEdit.setOnClickListener { listener.onClickEdit(product) }
            btnDelete.setOnClickListener { listener.onClickDelete(product) }
        }
    }

}
